File: src/go/ch01/1.4/l1/int_arith.go
package main

import "math/big"

// Exported bounds for consistency with demos/tests.
const (
	Int32Min int32 = -1 << 31
	Int32Max int32 = 1<<31 - 1
	Int64Min int64 = -1 << 63
	Int64Max int64 = 1<<63 - 1
)

/********** int32 **********/

func AddI32Checked(a, b int32) (int32, bool) {
	s := int64(a) + int64(b)
	if s < int64(Int32Min) || s > int64(Int32Max) {
		return 0, false
	}
	return int32(s), true
}

func AddI32Wrapping(a, b int32) int32 {
	return int32(uint32(a) + uint32(b)) // two's-complement wrap
}

func AddI32Saturating(a, b int32) int32 {
	s := int64(a) + int64(b)
	if s > int64(Int32Max) {
		return Int32Max
	}
	if s < int64(Int32Min) {
		return Int32Min
	}
	return int32(s)
}

func MulI32Checked(a, b int32) (int32, bool) {
	p := int64(a) * int64(b)
	if p < int64(Int32Min) || p > int64(Int32Max) {
		return 0, false
	}
	return int32(p), true
}

/********** int64 **********/

func AddI64Checked(a, b int64) (int64, bool) {
	if (b > 0 && a > Int64Max-b) || (b < 0 && a < Int64Min-b) {
		return 0, false
	}
	return a + b, true
}

func AddI64Wrapping(a, b int64) int64 {
	return int64(uint64(a) + uint64(b)) // explicit wrap
}

func AddI64Saturating(a, b int64) int64 {
	if b > 0 && a > Int64Max-b {
		return Int64Max
	}
	if b < 0 && a < Int64Min-b {
		return Int64Min
	}
	return a + b
}

func MulI64Checked(a, b int64) (int64, bool) {
	z := new(big.Int).Mul(big.NewInt(a), big.NewInt(b))
	if z.Cmp(big.NewInt(Int64Min)) < 0 || z.Cmp(big.NewInt(Int64Max)) > 0 {
		return 0, false
	}
	return z.Int64(), true
}

File: src/python/ch01/1.4/l1/int_arith.py
from typing import Tuple

INT32_MIN, INT32_MAX = -2_147_483_648, 2_147_483_647
INT64_MIN, INT64_MAX = -9_223_372_036_854_775_808, 9_223_372_036_854_775_807

# ---------- helpers to convert to signed from masked (two's complement) ----------
def _to_signed(masked: int, bits: int) -> int:
    sign = 1 << (bits - 1)
    full = (1 << bits)
    return masked - full if (masked & sign) else masked

# ---------- i32 ops ----------
def add_i32_checked(a: int, b: int) -> int:
    s = a + b
    if s < INT32_MIN or s > INT32_MAX:
        raise OverflowError("int32 overflow")
    return s

def add_i32_wrapping(a: int, b: int) -> int:
    s = (a + b) & 0xFFFFFFFF
    return _to_signed(s, 32)

def add_i32_saturating(a: int, b: int) -> int:
    s = a + b
    if s > INT32_MAX: return INT32_MAX
    if s < INT32_MIN: return INT32_MIN
    return s

def mul_i32_checked(a: int, b: int) -> int:
    # widen in Python, then bounds-check for i32 range
    p = a * b
    if p < INT32_MIN or p > INT32_MAX:
        raise OverflowError("int32 overflow (mul)")
    return p

# ---------- i64 ops ----------
def add_i64_checked(a: int, b: int) -> int:
    s = a + b
    if s < INT64_MIN or s > INT64_MAX:
        raise OverflowError("int64 overflow")
    return s

def add_i64_wrapping(a: int, b: int) -> int:
    s = (a + b) & 0xFFFFFFFFFFFFFFFF
    return _to_signed(s, 64)

def add_i64_saturating(a: int, b: int) -> int:
    s = a + b
    if s > INT64_MAX: return INT64_MAX
    if s < INT64_MIN: return INT64_MIN
    return s

def mul_i64_checked(a: int, b: int) -> int:
    p = a * b
    if p < INT64_MIN or p > INT64_MAX:
        raise OverflowError("int64 overflow (mul)")
    return p
